import/webview/plugin.cpp

```diff
@@ -74,6 +83,68 @@ void SailfishOSWebViewPlugin::initializeEngine(QQmlEngine *engine, const char *u
     webEngine->addObserver(QStringLiteral("clipboard:setdata"));
 }
 
+void SailfishOSWebViewPlugin::initUserAgentOverrides(const QString &path)
+{
+    // Read source json
+    QJsonDocument sourceDoc;
+    QFile sourceFile(MOZILLA_DATA_UA_UPDATE_SOURCE);
+    if (sourceFile.open(QIODevice::ReadOnly)) {
+        QByteArray line;
+        QByteArray sourceData;
+        while (!sourceFile.atEnd()) {
+            line = sourceFile.readLine();
+            if (!line.trimmed().startsWith("//"))
+                sourceData += line;
+        }
+        sourceFile.close();
+        QJsonParseError e;
+        sourceDoc = QJsonDocument::fromJson(sourceData, &e);
+        if (sourceDoc.isNull()) {
+            qWarning() << sourceFile.fileName() << "parse error" << e.errorString() << "at offset" << e.offset;
+            return;
+        }
+    } else {
+        qWarning() << "Could not open" << sourceFile.fileName();
+    }
+
+    // Make the mozilla dir if not there
+    QString mozillaDir = QString("%1/.mozilla/").arg(path);
+    QDir dir(mozillaDir);
+    if (!dir.exists())
+        dir.mkpath(dir.path());
+
+    // Read destination json
+    QJsonDocument destDoc;
+    QFile destFile(mozillaDir + MOZILLA_DATA_UA_UPDATE);
+    if (destFile.open(QIODevice::ReadWrite)) {
+        QJsonParseError e;
+        destDoc = QJsonDocument::fromJson(destFile.readAll(), &e);
+        if (destDoc.isNull()) {
+            qWarning() << destFile.fileName() << "parse error" << e.errorString() << "at offset" << e.offset;
+            // Not bailing out to get the corrupted file rewritten
+        }
+        QJsonObject source = sourceDoc.object();
+        QJsonObject dest = destDoc.object();
+        QStringList destKeys = dest.keys();
+        bool changed = false;
+        for (const QString &key : source.keys()) {
+            if (!destKeys.contains(key)) {
+                // Append missing key and value
+                dest.insert(key, source.value(key));
+                changed = true;
```

Comment:
Won't it still be the case that if a value is changed in `/usr/share/sailfish-browser/data/ua-update.json.in`, that it will be left unchanged by this?

Presumably the two should be kept synchronised at all times, but correct me if I'm misunderstanding. Originally I was thinking you could maybe just compare timestamps and refresh the entire file if ua-update.json.in was newer, but the approach you've used here should work too if it checks for value differences (rather than just existence of a key), I think.

Reply:
I don't know if webview downloads these overrides automatically from net like sailfish-browser does, so I didn't want to overwrite already existing values with possibly older data. Anyway safest bet is to use ua-update.json.in values so I'll change this.

Reply:
I hadn't thought of that. This seems to update nicely now at any rate.